### PR TITLE
fix: Extends nested fields to Events in /trackedentities endpoint [ NTI ] [ DHIS2-14407 ]

### PR DESCRIPTION
**Pull request by lucaCambi77**
With this PR we extend `TrackedEntityInstanceParams` to support nested fields for events with `EventParams`.
This way Tei, Enrollment, and Event have their own class for field -> param setup.
However, this can be verbose and I've created https://dhis2.atlassian.net/browse/TECH-1496 to further improve this configuration.
This PR also does a minor clean-up adding optional instead of null in case we can't find a relationship

### CODE REVIEW

**Review by teleivo — COMMENTED** [review]
Great @lucaCambi77 that you tackle this right away and that the `/tracker/events` endpoint can benefit from these changes 😄 !

It would be great if we can explore simplifying the design more. See my comment in EnrollmentEventsParams.

Other than that I only added some small suggestions.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/TrackedEntityFieldsParamMapper.java [hunk 1]
```diff
@@ -189,30 +192,56 @@ private static TrackedEntityInstanceParams withFieldEnrollmentAndRelationships(
...
+        List<String> pathToMatch )
     {
-        return fieldPaths.stream().filter( fp -> isEnrollmentField( fp, field ) && fp.isExclude() ).findFirst()
+        return fieldPaths.stream().filter( fp -> isNestedField( fp, nestedField, pathToMatch ) && fp.isExclude() )
+            .findFirst()
             .orElse( null );
     }
 
-    private static boolean isEnrollmentField( FieldPath path, String field )
+    private static boolean isNestedField( FieldPath fieldPath, String field, List<String> pathToMatch )
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/TrackedEntityFieldsParamMapper.java:242]: ok

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/EnrollmentEventsParams.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * @author Luca Camnbi
+ *
+ *         Class used to define inclusion in {@link EnrollmentParams} of
+ *         {@link EventParams} properties
+ */
+@With
+@Value
+public class EnrollmentEventsParams
```
[the remark this answers is not in the document]
Comment (lucaCambi77) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/EnrollmentEventsParams.java:41]: For now, I think it is ok. But as you are suggesting there is space for improvement that is why I've created https://dhis2.atlassian.net/browse/TECH-1496. We'll find a better design there

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/TrackedEntityFieldsParamMapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/fieldsmapper/TrackedEntityFieldsParamMapper.java]: You changed it to

```
    private static boolean isNestedField( FieldPath fieldPath, String field, List<String> fieldToMatch )
```

what I meant 😅 was

```java
    private static boolean isNestedField( FieldPath fieldPath, String fieldToMatch, List<String> pathToMatch )
```

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/TrackerTrackedEntitiesExportControllerTest.java [hunk 3]
```diff
@@ -475,29 +629,30 @@ private RelationshipType relationshipType( RelationshipEntity from, Relationship
         return type;
     }
 
-    private Relationship relationship( TrackedEntityInstance from, TrackedEntityInstance to )
+    private Relationship teiToTeiRelationship( TrackedEntityInstance from, TrackedEntityInstance to )
```
Comment (teleivo) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/TrackerTrackedEntitiesExportControllerTest.java:632]: It was intentional to write `relationship` as its clear what type `from` and `to` are. I think the method name changes make the tests less readable. Or was there some limitation that required this change 🤔 ?

**Review by teleivo — APPROVED** [review]

**Review by stian-sandvold — APPROVED** [review]

**Pull request merged**